beacon_node/eth2-libp2p/src/behaviour/handler/mod.rs

```diff
@@ -116,26 +140,40 @@ impl<TSpec: EthSpec> ProtocolsHandler for BehaviourHandler<TSpec> {
             Self::Error,
         >,
     > {
+        // TODO: is there any better place to check this?
+        if let KeepAlive::No = self.delegate.rpc().connection_keep_alive() {
+            if let KeepAlive::Yes = self.keep_alive {
+                // RPC inactivity timeout has passed
+                self.keep_alive = KeepAlive::Until(Instant::now() + self.shutdown_grace_time);
+                // TODO: should this be sent?
+                return Poll::Ready(ProtocolsHandlerEvent::Custom(BehaviourHandlerOut::IdleRPC));
+            }
+        }
+
+        // poll the delegate handler
+        // TODO: what do we keep sending to the behaviours if shutting down?
         match self.delegate.poll(cx) {
             Poll::Ready(ProtocolsHandlerEvent::Custom(event)) => {
                 return Poll::Ready(ProtocolsHandlerEvent::Custom(
                     BehaviourHandlerOut::Delegate(event),
                 ))
             }
             Poll::Ready(ProtocolsHandlerEvent::Close(err)) => {
-                return Poll::Ready(ProtocolsHandlerEvent::Close(err))
+                // TODO: one of the handlers has died, what here?
```

Comment:
If one handler wants to terminate the connection, lets terminate. As you've done here